components/calendar/src/astronomy.rs

```diff
@@ -198,12 +234,159 @@ impl Astronomical {
+
+        let mut result = angle;
+        let mut c_power = c;
+        for coef in list.into_iter().skip(1) {
+            result += coef * c_power;
+            c_power *= c;
+        }
+        result
+    }
+    // Calculates declination at a given Moment of UTC time for the latitude and longitude of an object lambda
+    pub(crate) fn declination(moment: Moment, beta: f64, lambda: f64) -> f64 {
+        let varepsilon = Self::obliquity(moment);
+
+        let sin_beta = sin_degrees(beta);
+        let sin_var = sin_degrees(varepsilon);
+        let cos_var = cos_degrees(varepsilon);
+        let cos_beta = cos_degrees(beta);
+        let sin_lambda = sin_degrees(lambda);
+
+        let add_1 = sin_beta * cos_var;
+        let add_2 = cos_beta * sin_var * sin_lambda;
+
+        let r = add_1 + add_2;
+
+        arcsin_degrees(
+            sin_degrees(beta) * cos_degrees(varepsilon)
+                + cos_degrees(beta) * sin_degrees(varepsilon) * sin_degrees(lambda),
+        )
+    }
+    pub(crate) fn right_ascension(
+        moment: Moment,
+        beta: f64,
+        lambda: f64,
+    ) -> Result<f64, &'static str> {
+        let varepsilon = Self::obliquity(moment);
+        arctan_degrees(
+            sin_degrees(lambda) * cos_degrees(varepsilon)
+                - tan_degrees(beta) * sin_degrees(varepsilon),
+            cos_degrees(lambda),
+        )
+    }
+
+    pub(crate) fn local_from_apparent(moment: Moment, location: Location) -> Moment {
+        moment - Self::equation_of_time(Location::universal_from_local(moment, location))
+    }
+
+    pub(crate) fn approx_moment_of_depression(
+        moment: Moment,
+        location: Location,
+        alpha: f64,
+        early: bool,
+    ) -> Result<Moment, CalendarError> {
+        let mut t = Self::sine_offset(moment, location, alpha);
+        let date = moment.as_rata_die().to_i64_date() as f64;
+        let alt = if alpha >= 0.0 {
+            if early {
+                date
+            } else {
+                date + 1.0
+            }
+        } else {
+            date + 12.0 / 24.0
+        };
+
+        let value = if t.abs() > 1.0 {
+            t = Self::sine_offset(Moment::new(alt), location, alpha);
+            t
+        } else {
+            t
+        };
+
+        if value.abs() <= 1.0 {
+            let offset = div_rem_euclid_f64(arcsin_degrees(value) / 360.0, 1.0 / 2.0).1;
+            let offset = if offset < -12.0 / 24.0 {
+                offset + 1.0
+            } else if offset >= 12.0 / 24.0 {
+                offset - 1.0
+            } else {
+                offset
+            };
+
+            let moment = Moment::new(
+                date + if early {
+                    6.0 / 24.0 - offset
+                } else {
+                    18.0 / 24.0 + offset
+                },
+            );
+            Ok(Self::local_from_apparent(moment, location))
+        } else {
+            return Err(CalendarError::DepressionAngleUnreachable);
```

Comment:
I think we can just use the error type `()` here, and convert in the calendar code caller. Or return an `OutOfRange` error or something. It really depends on how this is being called.